### DISCUSSION TITLE
how do we do releases?

### DISCUSSION THREAD

**Discussion by masinter** [Interlisp/medley#9] [configurations and releases]
github has a release system. Do we support lots of unix and mac lde ldex?
how much unix knowledge should we assume.
can people on windows machines run VirtualBox?
could we use a smaller distro than linux mint? (if we distribue a virtualbox VM)

should we use gitfs ? dropbox (read only) or distribute a ZIP of everything

**Comment by netdoll** [comment]
My suggestion for releases on GitHub: OVA/OVF VirtualBox appliance with everything set up (Mint is probably fine for your purposes, although if you want something more lightweight, there are a number of alternatives all the way down to boutique musl distributions like Alpine Linux), macOS releases in DMG (or tarballs, I'm assuming Apple does something with code signing here which might make it easier to just build from source), and tarballs for Linux (release source + build instructions for Debian-based distributions would be fine here).

For Unix knowledge, I think some simple build instructions (listing dependencies for Mint/Ubuntu/Debian etc.) would probably be sufficient, at least starting out.

I can also confirm that VirtualBox works fine from Windows, although it's maybe not the ideal environment for anything past simple experimentation.

W.R.T. the Dropbox, I'd keep it around but it's probably best to use GitHub's release system as soon as is practicable
~allison

**Comment by johnwcowan** [comment]
I would favor a Docker container over VBox, for these reasons:

1. Containers are more lightweight than VMs.

3. Docker for Windows is compatible with both Hyper/V (the MS virtualizer) and VBox, which are *not* compatible with each other.  Windows Server machines have a Docker server built in; for user machines, it's an easy download.

4. WSL2 is Microsoft's latest Linux personality for Windows: it uses a genuine Linux kernel and a very lightweight version of Hyper/V.  As such, it is compatible with Docker but not VBox.  (WSL1 is a different ball of wax: it is a Linux kernel *emulator* on top of the Windows Executive, the lower-level kernel.)

Note:  Decent X servers are freely available for both Windows (I like Cygwin/X) and Mac.

Of course, we can easily have both.  $EMPLOYER distributes its software, which runs in a container on AWS, to customers who don't trust the Internet or AWS with their precious bodily fluids^W^Wdocuments; they get the same container wrapped in a VM, which typically gets run at the client site on a VMWare box.

**Comment by masinter** [comment]
I don't quite understand how it would work with docker on windows -- Windows users would have to install an X server from somewhere else?  What would be in the distribution?
At least with VirtualBox I know I start with a clean image, login as linuxmint, sudo apt install a couple of things needed (X11, tcsh)
make a user LispUser, add it to the vboxsf group, share a folder, copy lde and ldex to /usr/local/bin, change the password of the linuxmint account, log in as LispUser ... and 10 more steps.  
But then a windows user would have to install virtualbox, download our virtual box vm, start it, and type medley at the prompt

linuxmint is pretty heavy; but any other one i'd have to find what packages were needed.

**Comment by johnwcowan** [comment]
Here are the generic instructions for a user on Windows, Mac, or Linux:

1. Download and run an X server if you don't have one already.  XQuartz is good for Mac, Xming or Cygwin/X for Windows. 
2. Download and run Docker if you don't have it already.
3. Get a free DockerHub ID (there's a web page to do it) if you don't have one already.
4. Open a terminal and type "docker login; docker run interlisp/medley.N", where N is a release number. 

This will download a Docker image off DockerHub (where we have strategically put it) and execute its startup program, which will be "medley".  Pretty soon Interlisp windows should start to pop up.  (Note that on Windows and Mac, X servers hide the root window by default, so the X windows are mixed in with native windows and are controlled by the native window manager.) 
Once the image is downloaded, it does not have to be downloaded again until there is a new release.

There's lots of information around the web on building Docker images; it's not particularly difficult.  You start with a pre-existing image of a stripped-down Linux system (there are many available) and add things as needed at build time.  When you are done, you upload the built image to DockerHub.  [Here's a tutorial.](https://www.mirantis.com/blog/how-do-i-create-a-new-docker-image-for-my-application/)

**Comment by masinter** [comment]
Interlisp doesn’t _use_  X. It just needs a bitmap that is shared with display. There is only one window, which should be run fullscreen. There are Medley features for running a shell in a Lisp window.

 

There are maiko #ifdefs that turn off X compilation.

It ran on DOS. 

Maybe there’s a way of making medley without X run in a container?

 

.  <https://github.com/notifications/beacon/AAIQTK3P5FLMULWWWBEJICTSAMAEVA5CNFSM4P2WVR6KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOFAPN7ZI.gif>

**Comment by johnwcowan** [comment]
Well, that makes things easier.  But what ran okay on a 640 x 480 DOS screen is going to look terrible on a 1080p (1920 x 1080), either too tiny or too blocky.

There is no problem having a container run without X; it's only if the program depends on X that you need an X server.

**Comment by Anzus** [comment]
![Screen Shot 2020-08-13 at 4 41 47 PM](https://user-images.githubusercontent.com/8584460/90194221-fdfea600-dd83-11ea-84fc-64616a82d109.jpg)

That's what it looks like in VirtualBox set to SVGA, scaled 200%. Looks like the DOS version does support 1024x768, but I'm not sure how to get VirtualBox to take that.

**Comment by masinter** [comment]
Just to test out, I did a "minimal" release of "Medley3.5.2-prealpha"

https://github.com/Interlisp/medley/releases 
Anyone who can run linux 386_64 should be able to load it.
The other files aren't strictly needed, are they?
